### ISSUE TITLE
[MENFORCER-435] Get rid of maven-dependency-tree dependency

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-enforcer#556]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** opened **[MENFORCER-435](https://issues.apache.org/jira/browse/MENFORCER-435?redirect=false)** and commented

As the m-enforcer-p depends on Maven 3.2.5 (since MENFORCER-419) all dependency resolutions should be done with Aether API directly instead of leveraging https://maven.apache.org/shared/maven-dependency-tree/. At the same time the error message should print the full path to the affected dependency instead of forcing users to use a dedicated call of `mvn dependency:tree` to locate the dependency (https://github.com/apache/maven-enforcer/blob/a06b47ba079b342d69a49d3cbad0fb546000f734/enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/AbstractBanDependencies.java#L111)

---

**Issue Links:**
- [MENFORCER-407](https://issues.apache.org/jira/browse/MENFORCER-407) Enforcer 3.0.0 breaks with Maven 3.8.4

- [MENFORCER-434](https://issues.apache.org/jira/browse/MENFORCER-434) Version 3.1.0 is not enforcing bannedDependencies rules

- [MENFORCER-437](https://issues.apache.org/jira/browse/MENFORCER-437) DependencyConvergence treats provided dependencies are runtime dependencies

- [MENFORCER-400](https://issues.apache.org/jira/browse/MENFORCER-400) DependencyConvergence - poor error message

- [MENFORCER-419](https://issues.apache.org/jira/browse/MENFORCER-419) Upgrade Maven to 3.2.5

- [MRESOLVER-277](https://issues.apache.org/jira/browse/MRESOLVER-277) --no-snapshot-update option does not prevent metadata.xml download
 (_**"is depended upon by"**_)

**Remote Links:**
- [GitHub Pull Request #198
](https://github.com/apache/maven-enforcer/pull/198)
- [GitHub Pull Request #202
](https://github.com/apache/maven-enforcer/pull/202)
- [GitHub Pull Request #214
](https://github.com/apache/maven-enforcer/pull/214)

**Backported to:** [3.2.0](https://github.com/apache/maven-enforcer/milestone/21?closed=1)

**Comment by jira-importer** [comment]
**[Andrzej J.](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ajarmoniuk)** commented

Is anyone working on this one already? If not, I'd gladly take it. I did a similar job for versions-maven-plugin recently.

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

Good luck [Andrzej J.](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ajarmoniuk) 

Assigned to me for tracking purpose

**Comment by jira-importer** [comment]
**[Andrzej J.](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ajarmoniuk)** commented

Oh I didn't know you were working on it. Anyway, since the required maven version is 3.2.5, it's safe to assume that we can use the new Resolver API and thus remove maven-compat cruft altogether. Here's my attempt from today (very much a WIP): https://github.com/apache/maven-enforcer/compare/master...ajarmoniuk:maven-enforcer:MENFORCER-435?expand=1

If you don't mind, I'll gladly continue your job.

**Comment by jira-importer** [comment]
**[Andrzej J.](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ajarmoniuk)** commented

Update: replaced maven-compat and dendency-tree cruft with resolver. All current unit and integration tests are green. Now polishing and adding traceability (parent) information where it's not present, so that no additional use of the dependency tree plugin is needed to track the packages causing validation errors.

**Comment by jira-importer** [comment]
**[Andrzej J.](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ajarmoniuk)** commented

Looks like this PR also resolves MENFORCER-434.

**Comment by jira-importer** [comment]
**[Andrzej J.](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ajarmoniuk)** commented

I have a question: as pointed out by [Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin), I should not modify `AbstractBanDependencies` as it might be an API some other custom rules may depend on. I thought that custom rules should only depend on the API defined in `enforcer-api`, like I did for `versions-maven-plugin`.

Apparently I was wrong.

Therefore, my question is: am I allowed to modify the interface of other classes, like e.g. DependencyVersionMap? I'd like to apply a `DependencySelector` to filter out excludes and includes already at the dependency node retrieval stage.

Thanks.
